runtime/drive_handler.go

```diff
@@ -0,0 +1,182 @@
+	return stubDriveHandler{
+		RootPath: path,
+	}
+}
+
+// StubDrivePaths will create stub drives and return the paths associated with
+// the stub drives.
+func (h *stubDriveHandler) StubDrivePaths(count int) ([]string, error) {
+	paths := []string{}
+	for i := 0; i < count; i++ {
+		driveID := fmt.Sprintf("stub%d", i)
+		path := filepath.Join(h.RootPath, driveID)
+
+		if err := h.createStubDrive(driveID, path); err != nil {
+			return nil, err
+		}
+
+		paths = append(paths, path)
+	}
+
+	return paths, nil
+}
+
+func (h *stubDriveHandler) createStubDrive(driveID, path string) error {
+	f, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			h.logger.WithError(err).Errorf("unexpected error during %v close", f.Name())
+		}
+	}()
+
+	stubContent, err := internal.GenerateStubContent(driveID)
+	if err != nil {
+		return err
+	}
+
+	if _, err := f.WriteString(stubContent); err != nil {
+		return err
+	}
+
+	info, err := f.Stat()
+	if err != nil {
+		return err
+	}
+
+	fileSize := info.Size()
+	sectorAmount := fileSize / fcSectorSize
+	truncateSize := fcSectorSize * sectorAmount
+
+	remainingBytes := fileSize % fcSectorSize
+	if remainingBytes != 0 {
+		// If there are any residual bytes, this means we've need to fill the
+		// appropriate sector size to ensure that the data is visible to
+		// Firecracker.
+		truncateSize += fcSectorSize
+	}
+
+	// Firecracker will not show any drives smaller than 512 bytes. In
+	// addition, the drive is read in chunks of 512 bytes; if the drive size is
+	// not a multiple of 512 bytes, then the remainder will not be visible to
+	// Firecracker. So we truncate to the appropriate size based on the residual
+	// bytes remaining.
+	if err := os.Truncate(path, truncateSize); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SetDrives will set the given drives and the offset to which the stub drives
+// start.
+func (h *stubDriveHandler) SetDrives(offset int64, d []models.Drive) {
+	h.drives = d
+	h.stubDriveIndex = offset
+}
+
+// GetDrives returns the associated stub drives
+func (h *stubDriveHandler) GetDrives() []models.Drive {
+	return h.drives
+}
+
+// InDriveSet will iterate through all the stub drives and see if the path
+// exists on any of the drives
+func (h *stubDriveHandler) InDriveSet(path string) bool {
+	for _, d := range h.GetDrives() {
+		if firecracker.StringValue(d.PathOnHost) == path {
+			return true
+		}
+	}
+
+	return false
+}
+
+// PatchStubDrive will replace the next available stub drive with the provided drive
+func (h *stubDriveHandler) PatchStubDrive(ctx context.Context, client firecracker.MachineIface, pathOnHost string) (*string, error) {
+	h.mutex.Lock()
+	defer func() {
+		h.stubDriveIndex++
+		h.mutex.Unlock()
+	}()
+
+	// Check to see if stubDriveIndex has increased more than the drive amount.
+	// This can occur when operations are racing on the atomic.LoadInt64
```

Comment:
At this point, is it actually due to a race, or is it simply a matter of the calling code trying to patch more drives than there are slots?